### ISSUE TITLE
Audit: write up new our plan of attack

### ISSUE THREAD

**Issue by SamStuckey** [department-of-veterans-affairs/va.gov-team#80624]

**Comment by SamStuckey** [comment]
At a high level we have 2 (non-new / in progress) categories of submissions in our database
- ones that are 'finalized' I.E. we are done with them because they got where they needed to be.
- ones that still require remediation


Our goal is to identify the latter by eliminating the former.  There are several ways to eliminate a successful submission:
- ones that went down the happy path. These will have a `submitted_claim_id`.
- ones that went down the backup path and were later transitioned to a success state insight Lighthouse.  These will have a `backup_submitted_claim_id` and a lighthouse status of `success` or `vbms`
- ones that were successfully 'batched' (aka soskinized)
  - These will appear in the attached CSV
  
There is one more category of submissions that for now we are hand waving over; anything that has been packaged by myself and sent to Emily for manual remediation (PDF packaging AKA the new failsafe workflow).  Some of these are in the state `in_remediation`, but we will need to double check to be sure we didn't miss anything.

**Comment by SamStuckey** [comment]
Here is a breakdown of how all instances of `Form526Submission` fit into our audit

[Uploading... Screenshot 2024-04-11 at 3.18.41 PM.png]

**Comment by emilytheis** [comment]
re: this line  _IF a misfit submission can be verified as successful via the VBMS side audit (Alex) consider marking it as such. (state TBD). We may or may not need to validate Alex's findings_ 

I think it's unlikely we'll get more information from the other audit that allows us to knock off submission IDs and say they were complete, but I don't have a strong opinion about taking it out here. 

The task about how exactly to identify duplicates and see if they're the same as a successful submission will probably need its own ticket, since it will be a modified version of your existing deduper script, but for the purposes of describing the initiative overall, this is good!

**Comment by SamStuckey** [comment]
UPDATE:
- all Form526Submission have been backfilled with their correct status.  
- All 526s have been extracted from the CSV. there is only ONE overlapping ID: 2663407 <- next step is figuring out which record type we should associate this with (`Form526Submission` or `SavedClaim::DependencyClaim`)

**Comment by SamStuckey** [comment]
UPDATE:
The attached CSV or IDs (some 686 and some 526) have been parsed by form type into these two lists:

[686_ids.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBM1Z0QkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--ff8b6aa8f9f9e5392abe534789f358c9f966f896/686_ids.txt)

[526_ids.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBM2R0QkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--9952752a6cc4f6772a0bd5bc5cf349cc84ff37f9/526_ids.txt)

**Comment by SamStuckey** [comment]
UPDATE:
- The attaches CSV of all 686 and 526 ids that were taken in to VBMS has been parsed into these two lists, by form type:

[686_ids.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNUZ0QkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--6ea24bd2e878eb3e88703a5815cd9fb424f5ec74/686_ids.txt)

[526_ids.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNUp0QkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--cc546e2edffcd37cbebcc34c90ba8b1c2ea99fdf/526_ids.txt)

**Comment by emilytheis** [comment]
We have a consideration to incorporate in the de-duping. If a submission is an exact duplicate, but the earliest possible exact duplicate is not the one remediated and came in on a different day, it's still a problem.

For example:

Submission `12335` sent Jan 20
exact duplicate submission `67890` sent Feb 2

If we remediated submission `6789` in the code yellow 1, the Veteran would still be missing out on possible payment. this will get very fussy, but sharing in case we think of a straightforward way to identify not only if it's an "untouched" submission is an exact duplicate, but if it's the earliest of the exact duplicates.
`

**Comment by SamStuckey** [comment]
UPDATE: marked all historical successfull primary path submissions as `delivered_to_primary`

**Comment by SamStuckey** [comment]
@emilytheis restating our desired outcome on the duplicates to sanity check:

## AXIOMS:
-  a duplicate can be said to be part of a duplicate 'set' (let's call it a dupe-set).
-  being a member of a dupe-set *does not* imply anything about order of creation, only that they are determined to be exact (or close-enough to be called exact, per our TBD rules to this effect) matches. 
-  An 'ignorable duplicate' is one where no remedial action is required for that specific record. we simply mark it as such and move on (exact marking mechanism TBD)

If the above is true then we can say
## Condition 1
IF a submission is part of a dupe-set 
  AND it is non-remediated
  AND there is a dupe-set member(s) that is both earlier (created_at) and already remediated
  THEN there is no remedial action to be taken on this record. This is an 'ignorable duplicate'
## Condition 2
IF a submission is part of a dupe-set 
  AND it is non-remediated
  AND there is *NO dupe-set member(s)* that is both earlier (created_at) and already remediated
  THEN this record will require remedial action

I think those two conditions catch all of the possible variations on the theme. Would you agree that they are correct?

NOTE that i think this complicates the using of 'state' for duplicate tagging.  Setting up a 'dupe set' type relationship is just as easy and almost certainly a better way to create a robust record of what's going on here.

**Comment by SamStuckey** [comment]
UPDATE: we cannot use state to track duplicates.  If two submissions are the same, because of the `created_at` problem you brought up, they still both need an independent concept of state.

So if we want to do something programatic with duplicates, we have a can of worms to dig through. I think it's worth a high level sanity check: @emilytheis what is our high level goal with duplicate identification?

**Comment by SamStuckey** [comment]
[Moving this duplicate problem to another ticket for discussion.](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/81350)

**Comment by SamStuckey** [comment]
TODO: roll in these failures as a sanity check. They may have already been remediated
- [file too large](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/79685)
- [impossible date](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/78931)

**Comment by SamStuckey** [comment]
UPDATE:

A list of non-deduped (probalby has duplicates in it) list of untouched submission IDs.

**Comment by SamStuckey** [comment]
UPDATE:

A list of non-deduped (probalby has duplicates in it) list of untouched submission IDs.

**Comment by SamStuckey** [comment]
UPDATE: my 'short list' of possible untouched submissions is done, with `118261` records.  Next steps are to
- develop an agreed upon deduping standard
- dedupe the list to get our result
- compare that list with VBMS

**Comment by SamStuckey** [comment]
[WIP document capturing the final stages of this work (deduping)](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/engineering_research/post_remediation_audit_for_untouched_submissions.md)

**Comment by SamStuckey** [comment]
UPDATE: Current short list:
[526_audit_shortlist.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeWVnQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--12da9c27bcdc395c61a2047a2e3237cd6014a982/wipn8923_526_audit_shortlist.txt)

There are 118261 records identified thus far as 'untouched'.  We can probably divide this be ~5 to approximate the number of duplicates based on what I've seen in the past.  That would leave us with about 23k untouched submissions, which is higher than I expected. 

Next step update:
- sanity check this list a bit.
- do the other stuff I mentioned above

**Comment by SamStuckey** [comment]
A few theories about these untouched
- it's possible these are lot's and lot's of duplicates, some even of happy path and backup path failures that were rejected by some uniqueness validation in the API
- it's possible many of these were started but never completed, e.g. still 'in progress'.  Will check this

**Comment by SamStuckey** [comment]
Here is the above file with any potentially in progress forms removed (118261 submissions)
[526_audit_short_list.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBK3FrQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--4313064e8b15d4f1ad10ef9b5c74426f68d97689/526_audit_short_list.txt)

**Comment by SamStuckey** [comment]
UPDATE: 
By applying the following agreed upon logic, we are able to create this MVP diffing algorythm
```
# - Given a set of duplicates
#   - Are any of them successful submissions?
#     - IF yes
#       - For every submission created after the success, mark as ignoreable
#       - For the submissions created before the success
#         - the earliest created_at is of intrest
#         - the others can be ignored
#     - IF no
#       - Grab the earliest one and mark the rest as duplicates
```

The result of for the first 10 uuids of interest is as follows:
```ruby
User UUID: 677cc319e1b44ff6b86388a24aff271f
Ignore:
	id: 279618
	created_at: 2020-07-27 16:34:09 UTC
	state: unprocessed
	id: 279693
	created_at: 2020-07-27 17:13:57 UTC
	state: unprocessed
Investigate:
	id: 279526
	created_at: 2020-07-27 15:40:27 UTC
	state: unprocessed


User UUID: 8ce0c653bab145c9b2b493214074d9af
Ignore:
	id: 279618
	created_at: 2020-07-27 16:34:09 UTC
	state: unprocessed
	id: 279693
	created_at: 2020-07-27 17:13:57 UTC
	state: unprocessed
Investigate:
	id: 279526
	created_at: 2020-07-27 15:40:27 UTC
	state: unprocessed


User UUID: 548b41cc912f402f88dfcdaba287849c
Ignore:
	id: 279618
	created_at: 2020-07-27 16:34:09 UTC
	state: unprocessed
	id: 279693
	created_at: 2020-07-27 17:13:57 UTC
	state: unprocessed
Investigate:
	id: 279526
	created_at: 2020-07-27 15:40:27 UTC
	state: unprocessed


User UUID: 1a4801f93be849c5a233cebd1e5cfa94

cannot process complex differences for: 1a4801f93be849c5a233cebd1e5cfa94

User UUID: 1a4801f93be849c5a233cebd1e5cfa94

cannot process complex differences for: 1a4801f93be849c5a233cebd1e5cfa94

User UUID: 1a4801f93be849c5a233cebd1e5cfa94

cannot process complex differences for: 1a4801f93be849c5a233cebd1e5cfa94

User UUID: 5eacc7787f30485ea896cf45cf725420

cannot process complex differences for: 5eacc7787f30485ea896cf45cf725420

User UUID: 5c38dcc24eff431abb510ff35effde25
Ignore:
	id: 279618
	created_at: 2020-07-27 16:34:09 UTC
	state: unprocessed
	id: 279693
	created_at: 2020-07-27 17:13:57 UTC
	state: unprocessed
Investigate:
	id: 279526
	created_at: 2020-07-27 15:40:27 UTC
	state: unprocessed


User UUID: c528d4027e86404d97b6c4ef6761c1b9
Ignore:
	id: 279618
	created_at: 2020-07-27 16:34:09 UTC
	state: unprocessed
	id: 279693
	created_at: 2020-07-27 17:13:57 UTC
	state: unprocessed
Investigate:
	id: 279526
	created_at: 2020-07-27 15:40:27 UTC
	state: unprocessed


User UUID: be9f5d085da14f30bbeb50af503466aa
Ignore:
	id: 279618
	created_at: 2020-07-27 16:34:09 UTC
	state: unprocessed
	id: 279693
	created_at: 2020-07-27 17:13:57 UTC
	state: unprocessed
Investigate:
	id: 279526
	created_at: 2020-07-27 15:40:27 UTC
	state: unprocessed
```

**Comment by SamStuckey** [comment]
UPDATE:
next step, per @emilytheis 
- generate a more verbose sample with actual value differences and share it via private VA channel

**Comment by SamStuckey** [comment]
UPDATE: 
Our current best guess is approximately 20k submissions across 5k vets that are 'untouched'.  This number is naive in that it ignores users with 'complex duplicates'.  [More on that here](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/engineering_research/complex_submission_deduplication.md)

Next Steps:
- Complete 'complex' deduping.
- Develop a sample set for review by stakeholders including
  - dupe report
  - full form content
  - created_at

**Comment by SamStuckey** [comment]
[document on Complex Deduping](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/engineering_research/complex_submission_deduplication.md)

**Comment by SamStuckey** [comment]
UPDATE: moving 'Develop sample set' to [this ticket](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/82576)

**Comment by SamStuckey** [comment]
UPDATE: the last lap here looks like
- finish complex deduping
- finish tagging
- get data rich sample
- finalize our failure count

**Comment by SamStuckey** [comment]
UPDATE: 
- [complex deduper is done and i've added some audit scripts here](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/tree/master/teams/benefits/scripts/526/untouched_submission_audit)

**Comment by SamStuckey** [comment]
UPDATE: tagging complete
unless I've made a logic error somewhere along the way, are counts are
- 21,241 untouched submissions
- 5,011 affected users.

**Comment by emilytheis** [comment]
One other item to add to the TODO is
- Export a list of the currently assumed untouched submission IDs (21k) so we can compare them to what is in Alex's shorter list.

**Comment by SamStuckey** [comment]
UPDATE: sanity checked my state tagging  this morning.  good results

- all submissions with a `submitted_claim_id` (happy path) have one of the following success statuses
  - ["delivered_to_primary", "processed_in_batch_remediation", "finalized_as_successful"]
- all submissions with a `backup_submitted_claim_id` have one of the following success states
  - ["finalized_as_successful", "processed_in_batch_remediation", "ignorable_duplicate", "delivered_to_backup"] 
- submissions lacking either path id have a mix of remediation and unprocess states; this is expected
 

 note that `finalized_as_successful` is a status given to backup submissions that receive a success state.  The reason some of our happy path submissions have that state is that, apparently, some submissions seem to have been submitted to *both* paths.  This is not good, but also not relevant to our investigation here.  These are successful and can be excluded.

TL;DR - submission tagging looks good. we can move forward using submission state as a solid source of truth for de-duping and indentification of untouched submissions.

**Comment by SamStuckey** [comment]
UPDATE:
- We updated the de-duping rules a few times and tags became unreliable.  however, It is easy to re-tag 'success' states, so i did that and just treated everything non-success as a potential failure
- using this data set and our updated deduper, we produced the following numbers
  - 8973 untouched submissions
  - 5680 affected users
 
Here is a list of these submission IDS:
[untouched_526_ids.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNW5hQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--0fedb0a81e25f22c8d33d04af51e41f4d5b67390/untouched_526_ids.txt)

A list of user uuids can be extrapolated from this list using
`Form526Submission.where(id: <ids>).pluck(:user_uuid).uniq`

TODO:
- finish our [funnel document](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/engineering_research/untouched_submission_audit/funnel_logic.md).  This will be important  if / when we need to explain these numbers
- Figure out what's wrong with tagging.  when manually retagging success I keep seeing a few 10s of submissions every week not getting their LH polling tag (finalized_as_success).  check on that job

**Comment by SamStuckey** [comment]
Marking as COMPLETE as we have a list of ids.  Follow up work on documentation and QA will be ticketed

**Comment by tblackwe** [comment]
Thank you @emilytheis. Kyle Matheny was able to get the server running again. I synced with Scott via slack and he will rerun the script using that file
